### PR TITLE
feat(versioning): support for adding, listing, and retrieving revisions

### PR DESCRIPTION
**Pull request by kulgan**
Added functions for:
* Getting the latest revision on an entry
* Adding a new revision to an entry
* Listing revisions for an entry

### CODE REVIEW

File: indexclient/client.py [hunk 1]
```diff
@@ -192,6 +192,61 @@ def create_alias(
         resp = self._put(url, headers=headers, data=data, auth=self.auth)
         return resp.json()
 
+    def get_latest_revision(self, did):
```
Comment (philloooo) [inline indexclient/client.py]: s/revision/version/g for all?

File: indexclient/client.py [hunk 2]
```diff
@@ -192,6 +192,61 @@ def create_alias(
...
+            * create a revision
+
+        Thr function should have the following signature
+            * func(str) -> str
+        The use of this function is to cater for cases where the version numbering is
+        complex
+        :type did: str
+        :param did: document id of the index to be revised
+        :type version_incr_func: function[str] -> str
+        :param version_incr_func: if not function is specified, it simply increments the latest version by 1
```
Comment (philloooo) [inline indexclient/client.py]: where does this happen?

File: tests/conftest.py [hunk 3]
```diff
@@ -1,9 +1,10 @@
 import pytest
+from cdisutilstest.code.conftest import indexd_server
+from indexclient.client import IndexClient
 
-from cdisutilstest.code.conftest import indexd_client, indexd_server
 
-@pytest.fixture(scope='function')
-def index_client(indexd_client):
+@pytest.fixture(scope='session')
```
Comment (jbarno) [inline tests/conftest.py]: Why are you session scoping a client? This means all modifications that a test makes are going to be visible to other tests.
- Which down the line, means mocks could potentially be dependent on the order of test execution
- cdisutilstest also exposes and indexd_client (but I kinda think it should live in here?)

Comment (jbarno) [reply tests/conftest.py]: Related PR from @philloooo : your not fully cleaning up indexd here:
https://github.com/uc-cdis/cdisutils-test/pull/12

File: tests/test_client.py [hunk 4]
```diff
@@ -6,18 +9,21 @@ def test_instantiate(index_client):
         size=5,
         urls=[]
     )
-    assert doc.size == 5
-    assert doc.hashes == hashes
+    if doc.size == 5 or doc.hashes == hashes:
```
Comment (jbarno) [inline tests/test_client.py]: why turn two lines into 5?

Comment (jbarno) [reply tests/test_client.py]: assert doc.size == 5 or doc.hashes == hashes if you want one or the other? -- but why?

File: tests/test_client.py [hunk 5]
```diff
@@ -26,3 +32,144 @@ def test_list_with_params(index_client):
...
+def test_get_latest_revision(index_client):
+    """
+    :type index_client: indexclient.client.IndexClient
+    """
+    hashes = {'md5': 'ab167e49d25b488939b1ede42752458c'}
+    doc = index_client.create(
+        hashes=hashes,
+        size=10,
+        file_name="removable.txt",
+        urls=["s3://service.hidden.us/lostinspace"]
```
Comment (jbarno) [inline tests/test_client.py]: :joy:

File: indexclient/client.py [hunk 6]
```diff
@@ -192,6 +192,61 @@ def create_alias(
...
 
+    def get_latest_revision(self, did):
+        # type: (str) -> Document | None
+        doc = self._get("index", did, "latest").json()
+
+        if doc and "did" in doc:
+            return Document(self, doc["did"], doc)
+        return None
+
+    def auto_add_revision(self, did, version_incr_func=None):
```
Comment (DrSpez) [inline indexclient/client.py]: Will do nothing if `version_incr_func` is not passed,
Better make += 1 to version by default?

File: tests/test_client.py [hunk 7]
```diff
@@ -26,3 +32,144 @@ def test_list_with_params(index_client):
...
+    else:
+        raise AssertionError()
+
+
+def test_invalid_input(index_client):
+    """
+    :type index_client: indexclient.client.IndexClient
+    """
+
+    with pytest.raises(HTTPError):
```
Comment (jbarno) [inline tests/test_client.py]: Great candidate for parametrization. Pass tuples (ExepectedException, latest_revision)

File: tests/test_client.py [hunk 8]
```diff
@@ -26,3 +32,144 @@ def test_list_with_params(index_client):
...
+    """
+    hashes = {'md5': 'ab167e49d25b488939b1ede42752458c'}
+    doc = index_client.create(
+        hashes=hashes,
+        size=10,
+        file_name="removable.txt",
+        urls=["s3://service.hidden.us/lostinspace"]
+    )
+    latest = index_client.get_latest_revision(doc.did)
+    if latest.did == doc.did or latest.file_name == doc.file_name or latest.hashes == doc.hashes:
```
Comment (jbarno) [inline tests/test_client.py]: Don't we want to ensure all of these are true?

File: indexclient/client.py (hunk 6 above)
Comment (jbarno) [reply indexclient/client.py]: :+1:  for standardization

File: indexclient/client.py (hunk 2 above)
Comment (kulgan) [reply indexclient/client.py]: When you want to update the version number whenever you add a new version of an entry and you don't want to change the fields. To be frank, even in GDC I don;t see this ever happening, so I'll yank it off

File: tests/conftest.py (hunk 3 above)
Comment (jbarno) [reply tests/conftest.py]: Right, usage of a shared client is the problem. If you are modifying what index client could return in one test a subsequent test *could* break. Additionally if one were to start mocking all requests and responses from IndexClient, the data returned would not be just for the test. Sharing the same instance of index_client is fine, but you must clean up all changes that the indexd_client made on a per test basis.

and cdisutilstest is ahead. you are cleaning 1/3 of the what you should be:
https://github.com/uc-cdis/cdisutils-test/pull/12/files#diff-2f85be2a3c761372f1109442ab7026f2R60

File: indexclient/client.py [hunk 9]
```diff
@@ -19,7 +19,7 @@ def handle_error(resp):
         try:
             json = resp.json()
             resp.reason = json["error"]
-        except:
+        except :
```
Comment (ProfOak) [inline indexclient/client.py]: 👀

Comment (kulgan) [reply indexclient/client.py]: lemme fix that

File: tests/conftest.py [hunk 10]
```diff
@@ -1,9 +1,10 @@
 import pytest
+from cdisutilstest.code.conftest import indexd_server
+from indexclient.client import IndexClient
 
-from cdisutilstest.code.conftest import indexd_client, indexd_server
```
Comment (philloooo) [inline tests/conftest.py]: just import the indexd_client fixture in cdisutilstest and use that, don't need to make another index_client?

File: tests/test_client.py [hunk 11]
```diff
@@ -26,3 +30,82 @@ def test_list_with_params(index_client):
             break
     if not found:
         raise AssertionError()
+
+
+def test_get_latest_revision(index_client):
```
Comment (philloooo) [inline tests/test_client.py]: revision -> version for all tests names?

File: indexclient/client.py [hunk 12]
```diff
@@ -192,6 +192,38 @@ def create_alias(
...
+        if doc and "did" in doc:
+            return Document(self, doc["did"], doc)
+        return None
+
+    def add_version(self, family_member_did, index_revision):
+        """
+
+        Args:
+            family_member_did (str): did of an existing index whose baseid will be shared
+            index_revision (Document): the document version to add to family
```
Comment (philloooo) [inline indexclient/client.py]: the arg names all seems weird, can we make them `current_did` and `new_doc`?

File: tests/conftest.py (hunk 10 above)
Comment (jbarno) [reply tests/conftest.py]: Definitely, a fixtures whole point is to not duplicate test code. Make index_client take indexd_client as an argument and return the indexd_client, while functionally scoped. Keeps indexclient's test code consistent, and whose to know how many different index clients may ever exist.
The cdisutils pin is 0.2.2 [here](https://github.com/uc-cdis/cdisutils-test/releases)
https://github.com/uc-cdis/cdisutils-test/blob/02d69e5e9be4305717674dcc9efb2055234d08e0/cdisutilstest/code/conftest.py#L34

Comment (kulgan) [reply tests/conftest.py]: @jbarno I made changes to index_client which is not on the one I will import, cos the one cdisutil does not have my changes.

**Pull request merged**